core/src/main/java/org/mule/runtime/core/routing/UntilSuccessful.java

```diff
@@ -38,183 +59,187 @@
   }
 
   @Override
-  public void start() throws MuleException {
-    super.start();
-    if (untilSuccessfulStrategy instanceof Startable) {
-      ((Startable) untilSuccessfulStrategy).start();
-    }
+  public void dispose() {
+    super.dispose();
+    timer.stop();
   }
 
   @Override
-  public void stop() throws MuleException {
-    if (untilSuccessfulStrategy instanceof Stoppable) {
-      ((Stoppable) untilSuccessfulStrategy).stop();
-    }
-    super.stop();
+  public Event process(Event event) throws MuleException {
+    return processToApply(event, this);
   }
 
   @Override
-  public boolean isMatch(final Event event, Event.Builder builder) throws MuleException {
-    return true;
+  public Publisher<Event> apply(Publisher<Event> publisher) {
+    if (isTransactionActive()) {
+      return from(publisher).flatMap(event -> {
+        try {
+          Exception lastExecutionException = null;
+          for (int i = 0; i <= maxRetries; i++) {
+            try {
+              Event result = processToApplyWithChildContext(event, nestedChain);
+              if (result == null) {
+                return empty();
+              }
+              if (shouldRetry.test(result)) {
+                throw new ExpressionFailureException(createStaticMessage(EXPRESSION_FAILED_MSG + event));
+              } else {
+                return just(result);
+              }
+            } catch (Exception e) {
+              if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Exception thrown inside until-successful ", e);
+              }
+              if (getRetryPredicate().test(e)) {
+                lastExecutionException = e;
+                if (i < maxRetries) {
+                  sleep(millisBetweenRetries);
+                }
+              } else {
+                return error(e);
+              }
+            }
+          }
+          return error(getThrowableFunction(event).apply(lastExecutionException));
+        } catch (Exception e) {
+          return error(e);
+        }
+      });
+    } else {
+      return from(publisher)
+          .flatMap(event -> Mono
+              .from(processWithChildContext(event, scheduleRoute(p -> Mono.from(p)
+                  .transform(nestedChain)
+                  .doOnNext(result -> {
+                    if (shouldRetry.test(result)) {
+                      throw new ExpressionFailureException(createStaticMessage(EXPRESSION_FAILED_MSG + event));
+                    }
+                  })), ofNullable(getLocation())))
+              .transform(p -> policyTemplate.applyPolicy(p, getRetryPredicate(), e -> {
+              }, getThrowableFunction(event))));
+    }
   }
 
-  @Override
-  protected Event route(final Event event) throws MuleException {
-    return untilSuccessfulStrategy.route(event, flowConstruct);
+  private Predicate<Throwable> getRetryPredicate() {
+    return e -> e instanceof ExpressionFailureException
```

Comment:
Because it's explicitly thrown above when failure expression fails, even though there is no error from the route. This exception is private. I will change name to `FailureExpressionAssertionException` to make it clearer.